sklearn/linear_model/tests/test_sgd.py

```diff
@@ -113,1102 +140,1181 @@ def decision_function(self, X, *args, **kw):
+        assert_array_equal(clf.predict(T), true_result)
 
-        clf = self.factory(alpha=0.01, learning_rate="constant",
-                           eta0=0.1, loss="squared_epsilon_insensitive")
-        clf.fit(X, Y)
-        assert_equal(1.0, np.mean(clf.predict(X) == Y))
 
-        clf = self.factory(alpha=0.01, loss="huber")
-        clf.fit(X, Y)
-        assert_equal(1.0, np.mean(clf.predict(X) == Y))
+@pytest.mark.parametrize('klass', [SGDClassifier, SparseSGDClassifier])
+def test_sgd_bad_l1_ratio(klass):
+    # Check whether expected ValueError on bad l1_ratio
+    assert_raises(ValueError, klass, l1_ratio=1.1)
 
-        clf = self.factory(alpha=0.01, learning_rate="constant", eta0=0.01,
-                           loss="squared_loss")
-        clf.fit(X, Y)
-        assert_equal(1.0, np.mean(clf.predict(X) == Y))
 
-    def test_warm_start_multiclass(self):
-        self._test_warm_start(X2, Y2, "optimal")
+@pytest.mark.parametrize('klass', [SGDClassifier, SparseSGDClassifier])
+def test_sgd_bad_learning_rate_schedule(klass):
+    # Check whether expected ValueError on bad learning_rate
+    assert_raises(ValueError, klass, learning_rate="<unknown>")
 
-    def test_multiple_fit(self):
-        # Test multiple calls of fit w/ different shaped inputs.
-        clf = self.factory(alpha=0.01, shuffle=False)
-        clf.fit(X, Y)
-        assert hasattr(clf, "coef_")
 
-        # Non-regression test: try fitting with a different label set.
-        y = [["ham", "spam"][i] for i in LabelEncoder().fit_transform(Y)]
-        clf.fit(X[:, :-1], y)
+@pytest.mark.parametrize('klass', [SGDClassifier, SparseSGDClassifier])
+def test_sgd_bad_eta0(klass):
+    # Check whether expected ValueError on bad eta0
+    assert_raises(ValueError, klass, eta0=0,
+                  learning_rate="constant")
 
 
-class SparseSGDClassifierTestCase(DenseSGDClassifierTestCase):
-    """Run exactly the same tests using the sparse representation variant"""
+@pytest.mark.parametrize('klass', [SGDClassifier, SparseSGDClassifier,
+                                   SGDRegressor, SparseSGDRegressor])
+def test_sgd_bad_alpha(klass):
+    # Check whether expected ValueError on bad alpha
+    assert_raises(ValueError, klass, alpha=-.1)
 
-    factory_class = SparseSGDClassifier
 
+@pytest.mark.parametrize('klass', [SGDClassifier, SparseSGDClassifier,
+                                   SGDRegressor, SparseSGDRegressor])
+def test_sgd_bad_penalty(klass):
+    # Check whether expected ValueError on bad penalty
+    assert_raises(ValueError, klass, penalty='foobar',
+                  l1_ratio=0.85)
 
-###############################################################################
-# Regression Test Case
 
-class DenseSGDRegressorTestCase(unittest.TestCase, CommonTest):
-    """Test suite for the dense representation variant of SGD"""
+@pytest.mark.parametrize('klass', [SGDClassifier, SparseSGDClassifier])
```

Comment:
you forgot SGD regressors
```
@pytest.mark.parametrize('klass', [SGDClassifier, SparseSGDClassifier,
                                   SGDRegressor, SparseSGDRegressor])
```

Reply:
Maybe move those 3 test cases that test both regressor as classifier a little bit up to above the "Classification Test Case" section title? (the diff is already impossible anyway, so don't need to care about that)